asmcomp/selectgen.ml

```diff
@@ -1193,9 +1214,20 @@ method emit_tail (env:environment) exp =
   | Ctuple _
   | Cexit _ ->
     self#emit_return env exp
-  | Cregion _ ->
-    (* FIXME tail calls out of regions *)
-    self#emit_return env exp
+  | Cregion e when env.region_tail ->
+     (* Region fusion *)
+     self#emit_tail env e
+  | Cregion e ->
+      let reg = self#regs_for typ_int in
+      self#insert env (Iop Ibeginregion) [| |] reg;
+      assert (env.regions = []);
+      let env = { env with regions = [reg]; region_tail = true } in
+      self#emit_tail env e
+  | Ctail e ->
+      if env.region_tail then
+        self#insert env (Iop Iendregion) (List.hd env.regions) [||];
+      self#emit_tail env' e
```

Comment:
I think `env'` is wrong here: you need to pop the region from `env.regions`, not just mark it as non-tail.

(I'm not sure this matters: `env.regions` is only used to compile `Cexit`, and I don't think that can cross `Ctail`)